src/Cardano/Wallet/Mnemonic.hs

```diff
@@ -0,0 +1,332 @@
+    deriving (Show, Typeable)
+
+
+data MnemonicError csz
+    = ErrMnemonicWords MnemonicWordsError
+    | ErrEntropy (EntropyError csz)
+    | ErrDictionary DictionaryError
+    | ErrForbidden
+    deriving (Show)
+
+
+
+-- | Smart-constructor for the Entropy
+mkEntropy
+    :: forall n csz. (ValidEntropySize n, ValidChecksumSize n csz)
+    => ByteString
+    -> Either (EntropyError csz) (Entropy n)
+mkEntropy = toEntropy
+
+
+ambiguousNatVal
+    :: forall n . (KnownNat n)
+    => Integer
+ambiguousNatVal = natVal @n Proxy
+
+-- | Generate Entropy of a given size using a random seed.
+--
+-- Example:
+--     do
+--       ent <- genEntropy :: IO (Entropy 12)
+genEntropy
+    :: forall n csz. (ValidEntropySize n, ValidChecksumSize n csz)
+    => IO (Entropy n)
+genEntropy =
+    let
+        size =
+            fromIntegral $ ambiguousNatVal @n
+        eitherToIO =
+            either (throwM . UnexpectedEntropyError) return
+    in
+        (eitherToIO . mkEntropy) =<< Crypto.getEntropy (size `div` 8)
+
+
+-- | Smart-constructor for the Mnemonic
+mkMnemonic
+    :: forall mw n csz.
+     ( ConsistentEntropy n mw csz
+     , EntropySize mw ~ n
+     )
+    => [Text]
+    -> Either (MnemonicError csz) (Mnemonic mw)
+mkMnemonic wordsm = do
+    phrase <- left ErrMnemonicWords
+        $ mnemonicPhrase @mw (toUtf8String <$> wordsm)
+
+    sentence <- left ErrDictionary
+        $ mnemonicPhraseToMnemonicSentence Dictionary.english phrase
+
+    entropy <- left ErrEntropy
+        $ wordsToEntropy sentence
+
+    when (isForbiddenMnemonic sentence) $ Left ErrForbidden
+
+    pure Mnemonic
+        { mnemonicToEntropy  = entropy
+        , mnemonicToSentence = sentence
+        }
+
+
+-- | Convert an Entropy to a corresponding Mnemonic Sentence
+entropyToMnemonic
+    :: forall mw n csz.
+     ( ValidMnemonicSentence mw
+     , ValidEntropySize n
+     , ValidChecksumSize n csz
+     , n ~ EntropySize mw
+     , mw ~ MnemonicWords n
+     )
+    => Entropy n
+    -> Mnemonic mw
+entropyToMnemonic entropy = Mnemonic
+    { mnemonicToSentence = entropyToWords entropy
+    , mnemonicToEntropy  = entropy
+    }
+
+
+-- | Convert 'Entropy' to a raw 'ByteString'
+entropyToByteString
+    :: Entropy n
+    -> ByteString
+entropyToByteString = entropyRaw
+
+
+-- | Constant-time comparison of any sentence with the 12-word example mnemonic
+isForbiddenMnemonic
+    :: (ValidMnemonicSentence mw)
+    => MnemonicSentence mw
+    -> Bool
+isForbiddenMnemonic sentence =
+    let
+        bytes =
+            sentenceToRawString sentence
+
+        forbiddenMnemonics = sentenceToRawString <$>
+            [ mnemonicToSentence (def @(Mnemonic 12))
+            ]
+    in
+        any (constEq bytes) forbiddenMnemonics
+
+
+sentenceToRawString
+    :: (ValidMnemonicSentence mw)
+    => MnemonicSentence mw
+    -> Basement.UArray Word8
+sentenceToRawString =
+    Basement.toBytes Basement.UTF8 . mnemonicSentenceToString Dictionary.english
+
+
+
```

Comment:
We don't need to leave so much vertical space between functions. One blank line is really enough! :)

Reply:
done